src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/TagHelperSemanticRangeVisitor.cs

```diff
@@ -312,35 +328,36 @@ private bool ClassifyTagName(MarkupTagHelperElementSyntax node)
                 throw new ArgumentNullException(nameof(node));
             }
 
+            var isElement = false;
             if (node.StartTag != null && node.StartTag.Name != null)
             {
-                var name = node.StartTag.Name.Content;
-
-                if (!HtmlFactsService.IsHtmlTagName(name))
-                {
-                    // We always classify non-HTML tag names as TagHelpers if they're within a MarkupTagHelperElementSyntax
-                    return true;
-                }
-
-                // This must be a well-known HTML tag name like 'input', 'br'.
-
                 var binding = node.TagHelperInfo.BindingResult;
                 foreach (var descriptor in binding.Descriptors)
                 {
-                    if (!descriptor.IsComponentTagHelper())
+                    if (!IsAttributeRules(descriptor.TagMatchingRules))
                     {
-                        return false;
+                        isElement = true;
                     }
                 }
 
-                if (name.Length > 0 && char.IsUpper(name[0]))
+                // Don't colorize if we're only a taghelper due to attributes
+                return isElement;
+            }
+
+            return false;
+        }
+
+        private bool IsAttributeRules(IEnumerable<TagMatchingRuleDescriptor> tagMatchingRules)
+        {
+            foreach(var rule in tagMatchingRules)
+            {
+                if (rule.TagName != "*")
```

Comment:
This isn't sufficient unfortunately. Not all TagMatchingRules on the descriptor are "active" meaning if a TagMatchingRule isn't applying to the current tag but another one is then you're screwed. For instance, imagine a TagHelper that had the following rules:

```
[HtmlTargetElement("markdown")]
[HtmlTargetElement("*", Attributes = "apply-markdown")]
```

With your current logic you'd highlight the element for something like `<foo apply-markdown />`.

It might be worht separating this bug fix from this PR to address it separately. It's a little more complex 😄